### PR TITLE
Make BruteForceSampler consider failed trials

### PR DESCRIPTION
**Pull request by Alnusjaponica**
## Motivation
Ensure consistent behaviors for failed trials of both`BruteForceSampler` and `GridSampler`.
The following is minimum reproducible code to make `BruteForceSampler` sample the same parameters multiple times.
```
import optuna


def objective(trial):
    x = trial.suggest_int("x", 0, 1)
    return [0, 1] if x else "a"


study = optuna.create_study(sampler=optuna.samplers.BruteForceSampler(seed=0))
study.optimize(objective, n_trials=2)

```

## Description of the changes
Consider trials in `TrialState.FAIL` as finished trials.

### CODE REVIEW

**Comment by contramundum53** [comment]
Shouldn't `BruteForceSampler` redo failed trials (by ignoring them)?

**Comment by Alnusjaponica** [comment]
I believe it should not redo failed trials, as long as `GridSampler` does not, for consistency. In addition, it is possible that the failure is due to misconfiguration of parameters and redoing those trials instead of unseen parameter setting is inefficient.

**Comment by toshihikoyanase** [comment]
@contramundum53 Could you continue the PR, please?
@not522 Could you join the review of this PR, please?

**Comment by contramundum53** [comment]
Hmm, a failed trial may result from a trial being preempted or the cloud computation resources becoming unstable. What about adding an option to specify whether failed trials should be skipped?

**Comment by not522** [comment]
> What about adding an option to specify whether failed trials should be skipped?

It sounds reasonable but may need discussion about the design. How about doing it with another PR?

File: tests/samplers_tests/test_brute_force.py [hunk 1]
```diff
@@ -275,3 +275,17 @@ def objective_decreasing_variable(trial: Trial) -> float:
...
+def test_study_optimize_with_failed_trials() -> None:
+    def objective(trial: Trial) -> float:
+        x = trial.suggest_int("x", 0, 99)  # NOQA[F811]
+        return np.nan
+
+    study = optuna.create_study(sampler=samplers.BruteForceSampler())
+    study.optimize(objective, n_trials=100)
+
+    expected_suggested_values = [{"x": i} for i in range(100)]
+    all_suggested_values = [t.params for t in study.trials]
```
Comment (not522) [inline tests/samplers_tests/test_brute_force.py:289]: Could you test the length of `all_suggested_values`?

**Review by contramundum53 — COMMENTED** [review]
The code LGTM except for @not522 's comment.

**Review by contramundum53 — APPROVED** [review]

**Review by not522 — APPROVED** [review]
LGTM!

**Pull request merged**